File: leetcode/js/watering-plants.js
/**
 * @param {number[]} plants
 * @param {number} capacity
 * @return {number}
 */
var wateringPlants = function (plants, capacity) {
  let result = 0;
  let completedIdx = 0;
  let curWater = capacity;
  let curIdx = -1;

  while (completedIdx < plants.length) {
    if (plants[completedIdx] <= curWater) {
      result += completedIdx - curIdx;
      curIdx = completedIdx;
      curWater -= plants[completedIdx];
      completedIdx++;
    } else {
      curIdx = -1;
      result += completedIdx;
      curWater = capacity;
    }
  }

  return result;
};

File: leetcode/rust/watering-plants.rs
impl Solution {
    pub fn watering_plants(plants: Vec<i32>, capacity: i32) -> i32 {
        let mut result = 0;
        let mut water = capacity;

        for i in 0..plants.len() {
            if water >= plants[i] {
                water -= plants[i];
                result += 1;
            } else {
                result += (2 * i) + 1;
                water = capacity - plants[i];
            }
        }

        result as i32
    }
}
